### PR TITLE
add trace support for kubelet csi-plugin

### PR DESCRIPTION
**Pull request by fatsheep9146**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

Add trace support for kubelet csi plugin

Fixes part of #113414

I used kind as local environment and use https://github.com/kubernetes-csi/csi-driver-host-path as demo csi driver. following is the span in demo

image

image

image

#### Does this PR introduce a user-facing change?

None

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: [<link>](https://github.com/kubernetes/enhancements/issues/2831)
- [Usage]: <link>
- [Other doc]: <link>
```

### CODE REVIEW

**Comment by fatsheep9146** [comment]
@dashpole could you help add label sig-instrumentation and review this pr?

File: pkg/kubelet/lifecycle/interfaces.go [hunk 1]
```diff
@@ -16,7 +16,7 @@ limitations under the License.
 
 package lifecycle
 
-import "k8s.io/api/core/v1"
+import v1 "k8s.io/api/core/v1"
```
Comment (dashpole) [inline pkg/kubelet/lifecycle/interfaces.go:19]: This appears unrelated.  Please revert.

File: pkg/kubelet/kubelet_test.go [hunk 2]
```diff
@@ -377,7 +377,7 @@ func newTestKubeletWithImageList(
 
 	var prober volume.DynamicPluginProber // TODO (#51147) inject mock
 	kubelet.volumePluginMgr, err =
-		NewInitializedVolumePluginMgr(kubelet, kubelet.secretManager, kubelet.configMapManager, token.NewManager(kubelet.kubeClient), allPlugins, prober)
+		NewInitializedVolumePluginMgr(kubelet, kubelet.secretManager, kubelet.configMapManager, token.NewManager(kubelet.kubeClient), allPlugins, prober, nil)
```
Comment (dashpole) [inline pkg/kubelet/kubelet_test.go:380]: Prefer using `go.opentelemetry.io/otel/trace.NewNoopTracerProvider()` to nil in tests.

File: pkg/volume/csi/csi_client.go [hunk 3]
```diff
@@ -529,19 +534,33 @@ func asSingleNodeMultiWriterCapableCSIAccessModeV1(am api.PersistentVolumeAccess
...
+	dialOpts = append(dialOpts,
 		grpc.WithAuthority("localhost"),
 		grpc.WithTransportCredentials(insecure.NewCredentials()),
 		grpc.WithContextDialer(func(ctx context.Context, target string) (net.Conn, error) {
 			return (&net.Dialer{}).DialContext(ctx, network, target)
 		}),
 		grpc.WithChainUnaryInterceptor(metricsManager.RecordMetricsInterceptor),
 	)
+
+	if tp != nil {
```
Comment (dashpole) [inline pkg/volume/csi/csi_client.go:551]: consider removing this check, and making sure Noops are passed to the CSI client

File: pkg/volume/csi/csi_client.go [hunk 4]
```diff
@@ -529,19 +534,33 @@ func asSingleNodeMultiWriterCapableCSIAccessModeV1(am api.PersistentVolumeAccess
...
+		tracingOpts := []otelgrpc.Option{
+			otelgrpc.WithPropagators(tracing.Propagators()),
+			otelgrpc.WithTracerProvider(tp),
+		}
+		// Even if there is no TracerProvider, the otelgrpc still handles context propagation.
+		// See https://github.com/open-telemetry/opentelemetry-go/tree/main/example/passthrough
+		dialOpts = append(dialOpts,
```
Comment (dashpole) [inline pkg/volume/csi/csi_client.go:558]: Should this be feature-gated?

File: pkg/volume/testing/volume_host.go [hunk 5]
```diff
@@ -250,6 +251,10 @@ func (f *fakeVolumeHost) WaitForKubeletErrNil() error {
 	})
 }
 
+func (f *fakeVolumeHost) GetTracerProvider() trace.TracerProvider {
+	return nil
```
Comment (dashpole) [inline pkg/volume/testing/volume_host.go:255]: Noop here as well

File: pkg/volume/csi/csi_plugin.go [hunk 6]
```diff
@@ -124,7 +124,7 @@ func (h *RegistrationHandler) RegisterPlugin(pluginName string, endpoint string,
 	})
 
 	// Get node info from the driver.
-	csi, err := newCsiDriverClient(csiDriverName(pluginName))
+	csi, err := newCsiDriverClient(csiDriverName(pluginName), nil)
```
Comment (dashpole) [inline pkg/volume/csi/csi_plugin.go:127]: Why is there no TracerProvider here?

File: pkg/controller/volume/persistentvolume/volume_host.go [hunk 7]
```diff
@@ -38,6 +39,10 @@ import (
 
 var _ vol.VolumeHost = &PersistentVolumeController{}
 
+func (ctrl *PersistentVolumeController) GetTracerProvider() trace.TracerProvider {
+	return nil
```
Comment (dashpole) [inline pkg/controller/volume/persistentvolume/volume_host.go:43]: Why don't we have a TracerProvider here?

File: pkg/controller/volume/expand/expand_controller.go [hunk 8]
```diff
@@ -370,6 +371,10 @@ func (expc *expandController) isNodeExpandComplete(logger klog.Logger, pvc *v1.P
 }
 
 // Implementing VolumeHost interface
+func (expc *expandController) GetTracerProvider() trace.TracerProvider {
+	return nil
```
Comment (dashpole) [inline pkg/controller/volume/expand/expand_controller.go:375]: Why is there no TracerProvider here?

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 9]
```diff
@@ -794,6 +795,10 @@ func (adc *attachDetachController) VolumeAttachmentLister() storagelistersv1.Vol
 // code to be implemented by all initializers even if the initializer does not
 // do mounting (like this attach/detach controller).
 // Issue kubernetes/kubernetes/issues/14217 to fix this.
+func (adc *attachDetachController) GetTracerProvider() trace.TracerProvider {
+	return nil
```
Comment (dashpole) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go:799]: Why is there no TracerProvider here?

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 10]
```diff
@@ -794,6 +795,10 @@ func (adc *attachDetachController) VolumeAttachmentLister() storagelistersv1.Vol
 // code to be implemented by all initializers even if the initializer does not
 // do mounting (like this attach/detach controller).
 // Issue kubernetes/kubernetes/issues/14217 to fix this.
```
Comment (dashpole) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go:797]: This comment appears to apply to GetPluginDir, not GetTracerProvider()

File: pkg/kubelet/kubelet_test.go (hunk 2 above)
Comment (fatsheep9146) [reply pkg/kubelet/kubelet_test.go:380]: Done

File: pkg/kubelet/lifecycle/interfaces.go (hunk 1 above)
Comment (fatsheep9146) [reply pkg/kubelet/lifecycle/interfaces.go:19]: Done

File: pkg/volume/csi/csi_client.go (hunk 3 above)
Comment (fatsheep9146) [reply pkg/volume/csi/csi_client.go:551]: after checking about the code, I think this check could be removed, since the tp is passed by volume_host in
https://github.com/kubernetes/kubernetes/blob/8323c71b0e0031da077b73c994fc74bae03c8bc7/pkg/kubelet/volume_host.go#L89

which is passed by function `NewInitializedVolumePluginMgr`, and the input param `traceprovider` in `NewInitializedVolumePluginMgr` is initialized in function `UnsecuredDependencies`

https://github.com/kubernetes/kubernetes/blob/8323c71b0e0031da077b73c994fc74bae03c8bc7/cmd/kubelet/app/server.go#L435-L441

which ensure the tp is `NoopTraceProvider` by default

File: pkg/volume/csi/csi_client.go (hunk 4 above)
Comment (fatsheep9146) [reply pkg/volume/csi/csi_client.go:558]: how about using the feature-gate `KubeletTracing`?

File: pkg/controller/volume/attachdetach/attach_detach_controller.go (hunk 9 above)
Comment (fatsheep9146) [reply pkg/controller/volume/attachdetach/attach_detach_controller.go:799]: The `attachdetach` package is used by `kube-controller-manager`, and since this pr is focusing on add trace for csi-plugin in `kubelet`, so I do not implement the `GetTracerProvider` interface used by `kube-controller-manager.`

File: pkg/controller/volume/expand/expand_controller.go (hunk 8 above)
Comment (fatsheep9146) [reply pkg/controller/volume/expand/expand_controller.go:375]: The `expand` package is used by `kube-controller-manager`, and since this pr is focusing on add trace for csi-plugin in `kubelet`, so I do not implement the `GetTracerProvider` interface used by `kube-controller-manager.`

File: pkg/controller/volume/persistentvolume/volume_host.go (hunk 7 above)
Comment (fatsheep9146) [reply pkg/controller/volume/persistentvolume/volume_host.go:43]: The `persistentvolume ` package is used by `kube-controller-manager`, and since this pr is focusing on add trace for csi-plugin in `kubelet`, so I do not implement the `GetTracerProvider` interface used by `kube-controller-manager.`

File: pkg/volume/testing/volume_host.go (hunk 5 above)
Comment (fatsheep9146) [reply pkg/volume/testing/volume_host.go:255]: Done

File: pkg/volume/csi/csi_client.go (hunk 4 above)
Comment (dashpole) [reply pkg/volume/csi/csi_client.go:558]: That sounds good.

File: pkg/volume/csi/csi_client.go (hunk 3 above)
Comment (dashpole) [reply pkg/volume/csi/csi_client.go:551]: exactly.  feel free to remove

File: pkg/controller/volume/persistentvolume/volume_host.go (hunk 7 above)
Comment (dashpole) [reply pkg/controller/volume/persistentvolume/volume_host.go:43]: got it, thanks.  Please use a Noop instead of nil

File: pkg/controller/volume/expand/expand_controller.go (hunk 8 above)
Comment (dashpole) [reply pkg/controller/volume/expand/expand_controller.go:375]: Please use a Noop instead of nil

File: pkg/controller/volume/attachdetach/attach_detach_controller.go (hunk 9 above)
Comment (dashpole) [reply pkg/controller/volume/attachdetach/attach_detach_controller.go:799]: Please use a Noop instead of nil

Comment (fatsheep9146) [reply pkg/controller/volume/attachdetach/attach_detach_controller.go:799]: Done

File: pkg/controller/volume/expand/expand_controller.go (hunk 8 above)
Comment (fatsheep9146) [reply pkg/controller/volume/expand/expand_controller.go:375]: Done

File: pkg/controller/volume/persistentvolume/volume_host.go (hunk 7 above)
Comment (fatsheep9146) [reply pkg/controller/volume/persistentvolume/volume_host.go:43]: Done

File: pkg/volume/csi/csi_client.go (hunk 3 above)
Comment (fatsheep9146) [reply pkg/volume/csi/csi_client.go:551]: Done

File: pkg/volume/csi/csi_client.go (hunk 4 above)
Comment (fatsheep9146) [reply pkg/volume/csi/csi_client.go:558]: Done

File: pkg/volume/csi/csi_plugin.go (hunk 6 above)
Comment (fatsheep9146) [reply pkg/volume/csi/csi_plugin.go:127]: I think since the function `RegisterPlugin` is  be used by kubelet's pluginmanager, which also managed the deviceplugin and draplugin. So if the function signature is changed (add traceprovider), the deviceplugin and draplugin may also should be changed.

File: pkg/kubelet/runonce_test.go [hunk 11]
```diff
@@ -103,7 +103,7 @@ func TestRunOnce(t *testing.T) {
 
 	plug := &volumetest.FakeVolumePlugin{PluginName: "fake", Host: nil}
 	kb.volumePluginMgr, err =
-		NewInitializedVolumePluginMgr(kb, fakeSecretManager, fakeConfigMapManager, nil, []volume.VolumePlugin{plug}, nil /* prober */)
+		NewInitializedVolumePluginMgr(kb, fakeSecretManager, fakeConfigMapManager, nil, []volume.VolumePlugin{plug}, nil /* prober */, nil)
```
Comment (dashpole) [inline pkg/kubelet/runonce_test.go:106]: Use a no-op here.

File: pkg/volume/csi/csi_plugin.go (hunk 6 above)
Comment (dashpole) [reply pkg/volume/csi/csi_plugin.go:127]: please use a no-op here, then.

**Review by dashpole — APPROVED** [review]

**Comment by bart0sh** [comment]
@fatsheep9146 If you still need this PR then please rebase, if not, please close the PR.

**Comment by fatsheep9146** [comment]
> @fatsheep9146 If you still need this PR then please rebase, if not, please close the PR.

I will rebase this ASAP.
